sdk/daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/UpgradeCheckMain.scala

```diff
@@ -26,13 +27,20 @@ case class UpgradeCheckMain(loggerFactory: NamedLoggerFactory) {
   implicit val ec: ExecutionContext = ExecutionContext.global
   implicit val loggingContext: LoggingContextWithTrace = LoggingContextWithTrace.empty
 
-  private def decodeDar(
+  private def decodeDarOrDalf(
       path: String
-  ): Either[CouldNotReadDar, Dar[(Ref.PackageId, Ast.Package)]] = {
+  ): Either[CouldNotReadDarOrDalf, Dar[(Ref.PackageId, Ast.Package)]] = {
     val s: String = s"Decoding DAR from ${path}"
     logger.debug(s)
     val result = DarDecoder.readArchiveFromFile(new File(path))
-    result.left.map(CouldNotReadDar(path, _))
+    result match {
+      case Left(darErr) =>
+        ArchiveDecoder.fromFile(new File(path)) match {
+          case Left(dalfErr) => Left(CouldNotReadDarOrDalf(path, darErr, dalfErr))
+          case Right(x) => Right(Dar(x, List()))
+        }
+      case Right(x) => Right(x)
+    }
   }
```

Comment:
Could we not have two decoding functions and two check function ?

Reply:
I personally prefer to be explicit, and have two different check functions.
That's being said if you want to have only one decode function, you should not use `Dar` as output type.

Reply:
I've reverted my changes to upgrade-check since Paul doesn't need them anymore, so I think we don't need to discuss this anymore - I'm still fond of the idea of accepting Dalfs, but we can revisit this if we need to later in another PR after 2.10.1 is cut.